File: app/src/main/java/net/imatruck/dsmanager/views/adapters/TasksArrayAdapter.java
package net.imatruck.dsmanager.views.adapters;

import android.content.Context;
import android.content.res.ColorStateList;
import android.support.annotation.NonNull;
import android.support.v4.content.res.ResourcesCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import net.imatruck.dsmanager.R;
import net.imatruck.dsmanager.models.Task;
import net.imatruck.dsmanager.utils.BytesFormatter;
import net.imatruck.dsmanager.utils.EtaFormatter;
import net.imatruck.dsmanager.utils.PercentFormatter;
import net.imatruck.dsmanager.utils.StatusFormatter;

import java.util.List;
import java.util.Locale;


public class TasksArrayAdapter extends RecyclerView.Adapter<TasksArrayAdapter.ViewHolder> {

    private Context context;
    private List<Task> tasks;
    private static TaskListOnClickListener itemClickListener;

    public TasksArrayAdapter(@NonNull Context context, @NonNull List<Task> tasks,
                             TaskListOnClickListener listener) {
        this.context = context;
        this.tasks = tasks;
        TasksArrayAdapter.itemClickListener = listener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(
                R.layout.tasks_list_row, parent, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Task task = tasks.get(position);

        if (task == null) {
            return;
        }

        String taskTitle = task.getTitle();
        int statusString = StatusFormatter.getStatusString(task.getStatus());
        String status = statusString != -1 ? context.getString(statusString) : task.getStatus();
        int colorRes = StatusFormatter.getStatusColor(task.getStatus());
        int color = ResourcesCompat.getColor(context.getResources(), colorRes, null);
        double size = task.getSize();
        double speedDownload = task.getAdditional().getTransfer().getSpeedDownload();
        double speedUpload = task.getAdditional().getTransfer().getSpeedUpload();
        double sizeDownloaded = task.getAdditional().getTransfer().getSizeDownloaded();
        double percent = PercentFormatter.toPercent(sizeDownloaded, size);

        holder.titleText.setText(taskTitle);
        holder.statusText.setText(status);
        holder.labelLeftText.setText(BytesFormatter.humanReadable(size, false));
        holder.labelRightText.setText(
                String.format(Locale.getDefault(), "↓ %s - ↑ %s - %s",
                        BytesFormatter.humanReadable(speedDownload, true),
                        BytesFormatter.humanReadable(speedUpload, true),
                        EtaFormatter.calculateEta(sizeDownloaded, size, speedDownload))
        );
        holder.progressBar.setProgress((int) percent);
        holder.progressBar.setProgressTintList(ColorStateList.valueOf(color));
        holder.progressText.setText(String.format(Locale.getDefault(), "%.0f%%", percent));
    }

    @Override
    public int getItemCount() {
        return tasks.size();
    }

    public Task getTask(int position) {
        return tasks.get(position);
    }

    public void clear() {
        tasks.clear();
    }

    public void addAll(List<Task> newTasks) {
        tasks.addAll(newTasks);
    }

    public boolean isEmpty() {
        return tasks.isEmpty();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView titleText;
        TextView statusText;
        ProgressBar progressBar;
        TextView progressText;
        TextView labelLeftText;

        TextView labelRightText;
        ViewHolder(View itemView) {
            super(itemView);
            titleText = (TextView) itemView.findViewById(R.id.row_task_title);
            statusText = (TextView) itemView.findViewById(R.id.row_task_status);
            progressBar = (ProgressBar) itemView.findViewById(R.id.row_task_progress_bar);
            progressText = (TextView) itemView.findViewById(R.id.row_task_progress_text);
            labelLeftText = (TextView) itemView.findViewById(R.id.row_task_left_label);
            labelRightText = (TextView) itemView.findViewById(R.id.row_task_right_label);

            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            itemClickListener.onItemClick(this.getLayoutPosition());
        }
    }
}

File: app/src/main/java/net/imatruck/dsmanager/views/adapters/TasksArrayAdapter.kt
package net.imatruck.dsmanager.views.adapters

import android.content.Context
import android.content.res.ColorStateList
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import net.imatruck.dsmanager.R
import net.imatruck.dsmanager.models.Task
import net.imatruck.dsmanager.utils.BytesFormatter
import net.imatruck.dsmanager.utils.EtaFormatter
import net.imatruck.dsmanager.utils.PercentFormatter
import net.imatruck.dsmanager.utils.StatusFormatter
import java.util.*


class TasksArrayAdapter(private val context: Context, private val tasks: MutableList<Task>,
                        listener: TaskListOnClickListener) : RecyclerView.Adapter<TasksArrayAdapter.ViewHolder>() {

    init {
        TasksArrayAdapter.itemClickListener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(
                R.layout.tasks_list_row, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val task = tasks[position]

        val taskTitle = task.title
        val statusString = StatusFormatter.getStatusString(task.status)
        val status = if (statusString != -1) context.getString(statusString) else task.status
        val colorRes = StatusFormatter.getStatusColor(task.status)
        val color = ResourcesCompat.getColor(context.resources, colorRes, null)
        val size = task.size
        val speedDownload = task.additional!!.transfer!!.speedDownload
        val speedUpload = task.additional.transfer!!.speedUpload
        val sizeDownloaded = task.additional.transfer.sizeDownloaded
        val percent = PercentFormatter.toPercent(sizeDownloaded, size)

        holder.titleText.text = taskTitle
        holder.statusText.text = status
        holder.labelLeftText.text = BytesFormatter.humanReadable(size, false)
        holder.labelRightText.text = String.format(Locale.getDefault(),
                "↓ ${BytesFormatter.humanReadable(speedDownload, true)} - " +
                        "↑ ${BytesFormatter.humanReadable(speedUpload, true)} - " +
                        EtaFormatter.calculateEta(sizeDownloaded, size, speedDownload))
        holder.progressBar.progress = percent.toInt()
        holder.progressBar.progressTintList = ColorStateList.valueOf(color)
        holder.progressText.text = String.format(Locale.getDefault(), "%.0f%%", percent)
    }

    override fun getItemCount(): Int {
        return tasks.size
    }

    fun getTask(position: Int): Task {
        return tasks[position]
    }

    fun clear() {
        tasks.clear()
    }

    fun addAll(newTasks: List<Task>) {
        tasks.addAll(newTasks)
    }

    val isEmpty: Boolean
        get() = tasks.isEmpty()

    class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        internal var titleText: TextView = itemView.findViewById(R.id.row_task_title) as TextView
        internal var statusText: TextView = itemView.findViewById(R.id.row_task_status) as TextView
        internal var progressBar: ProgressBar = itemView.findViewById(R.id.row_task_progress_bar) as ProgressBar
        internal var progressText: TextView = itemView.findViewById(R.id.row_task_progress_text) as TextView
        internal var labelLeftText: TextView = itemView.findViewById(R.id.row_task_left_label) as TextView
        internal var labelRightText: TextView = itemView.findViewById(R.id.row_task_right_label) as TextView

        init {
            itemView.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            itemClickListener!!.onItemClick(this.layoutPosition)
        }
    }

    companion object {
        private var itemClickListener: TaskListOnClickListener? = null
    }
}
